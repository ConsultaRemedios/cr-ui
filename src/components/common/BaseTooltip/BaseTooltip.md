```vue
const tooltip = {
  top: false,
  bottom: false,
  left: false,
  right: false,
  dark: false
};

<div>
  <BaseButton @click="tooltip.top = !tooltip.top">
    Tooltip Top

    <BaseTooltip :show="tooltip.top" width="150px">
      Tooltip Example
    </BaseTooltip>
  </BaseButton>

  <BaseButton @click="tooltip.bottom = !tooltip.bottom">
    Tooltip Bottom

    <BaseTooltip :show="tooltip.bottom" position="bottom">
      Tooltip Example
    </BaseTooltip>
  </BaseButton>

  <BaseButton @click="tooltip.left = !tooltip.left">
    Tooltip Left

    <BaseTooltip :show="tooltip.left" position="left">
      Tooltip Example
    </BaseTooltip>
  </BaseButton>

  <BaseButton @click="tooltip.right = !tooltip.right">
    Tooltip Right

    <BaseTooltip :show="tooltip.right" position="right">
      Tooltip Example
    </BaseTooltip>
  </BaseButton>

  <BaseButton @click="tooltip.dark = !tooltip.dark">
    Tooltip Dark

    <BaseTooltip :show="tooltip.dark" position="bottom" theme="dark">
      Tooltip Example
    </BaseTooltip>
  </BaseButton>
</div>
```
