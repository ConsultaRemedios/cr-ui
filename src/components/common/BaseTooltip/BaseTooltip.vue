<template>
  <transition
    :enter-class="$style.enter"
    :enter-active-class="$style.enterActive"
    :enter-to-class="$style.enterTo"
    :leave-class="$style.leave"
    :leave-active-class="$style.leaveActive"
    :leave-to-class="$style.leaveTo"
  >
    <div
      v-show="show"
      :class="[$style.wrapper, themeClass]"
      :style="style"
    >
      <slot></slot>

      <div :class="$style.arrow" x-arrow></div>
    </div>
  </transition>
</template>

<script>
  import Popper from 'popper.js';

  export default {
    name: 'BaseTooltip',
    props: {
      show: {
        type: Boolean,
        required: true,
      },

      position: {
        type: String,
        default: 'top',
      },

      fixed: {
        type: Boolean,
        default: false,
      },

      width: {
        type: String,
        default: 'auto',
      },

      height: {
        type: String,
        default: 'auto',
      },

      theme: {
        type: String,
        default: 'light',
      },
    },

    data() {
      return {
        positionFitted: this.position,
        style: {
          width: this.width,
          height: this.height,
        },
        reference: null,
        content: null,
      };
    },

    mounted() {
      this.reference = this.$el.parentNode || document.body;
      this.content = this.$el;

      document.body.appendChild(this.content);

      this.initPopper();
    },

    destroyed() {
      this.destroyPopper();

      this.reference.appendChild(this.content);
    },

    methods: {
      initPopper() {
        this.popper = new Popper(this.reference, this.content, {
          placement: this.position,
          removeOnDestroy: true,
          positionFixed: this.fixed,
        });
      },

      updatePopper() {
        if (this.popper) {
          this.popper.update();
        }
      },

      destroyPopper() {
        if (this.popper) {
          this.popper.destroy();
          this.popper = null;
        }
      },
    },

    computed: {
      themeClass() {
        if (this.theme === 'dark') return this.$style.dark;
        return this.$style.light;
      },
    },

    watch: {
      show() {
        if (!this.show) return;

        this.$nextTick(() => {
          this.updatePopper();
        });
      },
    },
  };
</script>

<style module>
  .light {
    border: 1px solid #DADADA;
    color: #333333;
    background-color: #FFFFFF;
  }

  .dark {
    color: #FFFFFF;
    background-color: #333333;
    border-radius: 2px;
  }

  .wrapper {
    position: absolute;
    box-shadow: 0 0 6px 0 rgba(0,0,0,0.31);
    z-index: 9999;
    padding: 15px;
    opacity: 1;
  }

  .wrapper[x-placement^="top"] {
    margin-bottom: 15px;
  }

  .wrapper[x-placement^="bottom"] {
    margin-top: 15px;
  }

  .wrapper[x-placement^="left"] {
    margin-right: 15px;
  }

  .wrapper[x-placement^="right"] {
    margin-left: 15px;
  }

  .arrow {
    position: absolute;
    overflow: hidden;
  }

  .light .arrow::after {
    background: #FFF;
  }

  .dark .arrow::after {
    background: #333333;
  }

  .arrow::after {
    content: '';
    position: absolute;
    width: 14px;
    height: 14px;
    transform: rotate(45deg);
    box-shadow: 0 0 6px 0 rgba(0,0,0,0.31);
  }

  .wrapper[x-placement^="top"] .arrow {
    width: 30px;
    height: 15px;
    top: 100%;
    left: 50%;
    transform: translateX(-50%);
    margin-left: 30px;
  }

  .wrapper[x-placement^="top"] .arrow::after {
    bottom: 8px;
    left: 8px;
  }

  .wrapper[x-placement^="left"] .arrow {
    width: 15px;
    height: 30px;
    left: 100%;
    top: 50%;
    transform: translateY(-50%);
    margin-top: 25px;
  }

  .wrapper[x-placement^="left"] .arrow::after {
    top: 8px;
    right: 8px;
  }

  .wrapper[x-placement^="right"] .arrow {
    width: 15px;
    height: 30px;
    right: 100%;
    top: 50%;
    transform: translateY(-50%);
    margin-top: 25px;
  }

  .wrapper[x-placement^="right"] .arrow::after {
    top: 8px;
    left: 8px;
  }

  .wrapper[x-placement^="bottom"] .arrow {
    width: 30px;
    height: 15px;
    bottom: 100%;
    left: 50%;
    transform: translateX(-50%);
    margin-left: 30px;
  }

  .wrapper[x-placement^="bottom"] .arrow::after {
    top: 8px;
    left: 8px;
  }

  .enterActive,
  .leaveActive {
    transition: opacity 0.2s ease-out;
  }

  .enter,
  .leaveTo {
    opacity: 0;
  }

  .enterTo,
  .leave {
    opacity: 1;
  }
</style>
