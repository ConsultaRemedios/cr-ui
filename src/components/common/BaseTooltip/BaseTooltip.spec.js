import { mount } from '@vue/test-utils';
import PopperJS from 'popper.js';
import Vue from 'vue';
import BaseTooltip from './BaseTooltip.vue';

jest.mock('popper.js', () => {
  const PopperJS = jest.requireActual('popper.js');

  const Popper = jest.fn(() => {
    return {
      destroy: jest.fn(),
      update: jest.fn(),
      scheduleUpdate: jest.fn(),
    };
  });

  Popper.placements = PopperJS.placements;

  return Popper;
});

describe('BaseTooltip', () => {
  let wrapper;

  const mountWithParent = (opts = {}) => {
    const { props } = opts;

    const defaultProps = {
      show: true,
    };

    const Component = Vue.component('with-parent', {
      render(h) {
        return h('div', { class: 'parent' }, [
          h(BaseTooltip, {
            props: { defaultProps, ...props },
          }, [
            'Tooltip Example'
          ])
        ]);
      }
    });

    return mount(Component).find(BaseTooltip);
  };

  const mountWithRect = (position = 'top', fixed = false) => {
    const localWrapper = mount(BaseTooltip, {
      propsData: { position, fixed, show: true }
    });

    localWrapper.vm.reference.getBoundingClientRect = jest.fn(() => {
      return { y: 100, x: 100, height: 60, width: 200 };
    });

    localWrapper.vm.content.getBoundingClientRect = jest.fn(() => {
      return { height: 40, width: 60 };
    });

    localWrapper.vm.$mount();

    return localWrapper;
  };

  beforeAll(() => {
    Object.defineProperty(document.body, 'offsetHeight', {
      get() { return 800; },
    });

    window.scrollY = 0;
    window.innerWidth = 600;
  });

  afterEach(() => {
    wrapper.destroy();
    jest.clearAllTimers();
  });

  describe('Lifecycle', () => {
    describe('#mounted', () => {
      describe('Instance variables', () => {
        describe('When tooltip has no parent', () => {
          it('assigns body as reference element', () => {
            wrapper = mountWithRect();
            expect(wrapper.vm.reference).toEqual(document.body);
          });
        });

        describe('When tooltip has parent', () => {
          it('assigns it as reference element', () => {
            wrapper = mountWithParent();
            expect(wrapper.vm.reference.className).toEqual('parent');
          });
        });
      });

      it('moves component element to the end of body', () => {
        wrapper = mountWithParent();
        expect(wrapper.vm.$el).toEqual(document.body.lastElementChild);
      });

      it('calls initPopper method', () => {
        const initPopper = jest.fn();

        wrapper = mount(BaseTooltip, {
          propsData: { show: true },
          methods: { initPopper }
        });

        expect(initPopper).toHaveBeenCalledTimes(1);
      });
    });

    describe('#destroyed', () => {
      it('calls destroyPopper method', () => {
        const destroyPopper = jest.fn();

        wrapper = mount(BaseTooltip, {
          propsData: { show: true },
          methods: { destroyPopper }
        });
        wrapper.destroy();

        expect(destroyPopper).toHaveBeenCalledTimes(1);
      });

      it('moves element from body to reference element', () => {
        wrapper = mount(BaseTooltip, {
          propsData: { show: true }
        });

        jest.spyOn(wrapper.vm.reference, 'appendChild');
        wrapper.destroy();

        expect(wrapper.vm.reference.appendChild).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.reference.appendChild).toHaveBeenCalledWith(wrapper.vm.$el);
      })
    });
  });

  describe('Methods', () => {
    describe('#initPopper', () => {
      it('creates a Popper', () => {
        wrapper = mountWithRect();
        wrapper.vm.initPopper();

        expect(wrapper.vm.popper).not.toBeNull()
      });

      it('instantiates popper properly when "fixed" prop is "false"', () => {
        wrapper = mountWithRect();
        wrapper.vm.initPopper();

        expect(PopperJS).toHaveBeenCalledWith(
          wrapper.vm.reference,
          wrapper.vm.content,
          {
            placement: 'top',
            removeOnDestroy: true,
            positionFixed: false,
            modifiers: {
              preventOverflow: {
                enabled: true,
                boundariesElement: 'window',
              },
            },
          }
        );
      });

      it('instantiates popper properly when "fixed" prop is "true"', () => {
        wrapper = mountWithRect('top', true);
        wrapper.vm.initPopper();

        expect(PopperJS).toHaveBeenCalledWith(
          wrapper.vm.reference,
          wrapper.vm.content,
          {
            placement: 'top',
            removeOnDestroy: true,
            positionFixed: false,
            modifiers: {
              preventOverflow: {
                enabled: true,
                boundariesElement: 'window',
              },
            },
          }
        );
      });
    });

    describe('#updatePopper', () => {
      describe('When popper is already created', () => {
        it('calls "update" method from Popper', () => {
          wrapper = mountWithRect();
          wrapper.vm.initPopper();
          wrapper.vm.updatePopper();

          expect(wrapper.vm.popper.update).toHaveBeenCalled();
        });
      });
    });

    describe('#destroyPopper', () => {
      it('destroy a Popper', () => {
        wrapper = mountWithRect();
        wrapper.vm.initPopper();
        wrapper.vm.destroyPopper();

        expect(wrapper.vm.popper).toBeNull()
      });
    });
  });

  describe('Watchers', () => {
    describe('#show', () => {
      describe('When show is true', () => {
        it('calls updatePopper', async () => {
          const updatePopper = jest.fn();

          wrapper = mount(BaseTooltip, {
            propsData: { show: false },
            methods: { updatePopper }
          });

          updatePopper.mockClear();
          wrapper.setProps({ show: true });

          await wrapper.vm.$nextTick();

          expect(updatePopper).toHaveBeenCalledTimes(1);
        });
      });

      describe('When show is false', () => {
        it('does not call updatePopper', (done) => {
          const updatePopper = jest.fn();

          wrapper = mount(BaseTooltip, {
            propsData: { show: true },
            methods: { updatePopper }
          });

          updatePopper.mockClear();
          wrapper.setProps({ show: false });

          wrapper.vm.$nextTick(() => {
            expect(updatePopper).not.toHaveBeenCalled();
            done();
          });
        });
      });
    });
  });

  describe('Snapshots', () => {
    const mountForSnapshot = (position, show = true, theme = 'light') => {
      return mount(BaseTooltip, {
        propsData: { position, show, theme },
        slots: {
          default: 'Example tooltip text'
        }
      });
    }

    it('matches Snapshot when show is false', () => {
      expect(mountForSnapshot('top', false).element).toMatchSnapshot();
    });

    it('matches Snapshot when position is top', () => {
      expect(mountForSnapshot().element).toMatchSnapshot();
    });

    it('matches Snapshot when position is bottom', () => {
      expect(mountForSnapshot('bottom').element).toMatchSnapshot();
    });

    it('matches Snapshot when position is left', () => {
      expect(mountForSnapshot('left').element).toMatchSnapshot();
    });

    it('matches Snapshot when position is right', () => {
      expect(mountForSnapshot('right').element).toMatchSnapshot();
    });

    it('matches Snapshot when theme is dark', () => {
      expect(mountForSnapshot('right', true, 'dark').element).toMatchSnapshot();
    });
  });
});
