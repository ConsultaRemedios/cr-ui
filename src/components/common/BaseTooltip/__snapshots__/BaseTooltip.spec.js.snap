// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTooltip Snapshots matches Snapshot when position is bottom 1`] = `
<transition-stub
  enteractiveclass="enterActive"
  enterclass="enter"
  entertoclass="enterTo"
  leaveactiveclass="leaveActive"
  leaveclass="leave"
  leavetoclass="leaveTo"
>
  <div
    class="wrapper light"
    style="width: auto; height: auto;"
  >
    Example tooltip text 
    <div
      class="arrow"
      data-popper-arrow=""
    />
  </div>
</transition-stub>
`;

exports[`BaseTooltip Snapshots matches Snapshot when position is left 1`] = `
<transition-stub
  enteractiveclass="enterActive"
  enterclass="enter"
  entertoclass="enterTo"
  leaveactiveclass="leaveActive"
  leaveclass="leave"
  leavetoclass="leaveTo"
>
  <div
    class="wrapper light"
    style="width: auto; height: auto;"
  >
    Example tooltip text 
    <div
      class="arrow"
      data-popper-arrow=""
    />
  </div>
</transition-stub>
`;

exports[`BaseTooltip Snapshots matches Snapshot when position is not passed 1`] = `
<transition-stub
  enteractiveclass="enterActive"
  enterclass="enter"
  entertoclass="enterTo"
  leaveactiveclass="leaveActive"
  leaveclass="leave"
  leavetoclass="leaveTo"
>
  <div
    class="wrapper light"
    style="width: auto; height: auto;"
  >
    Example tooltip text 
    <div
      class="arrow"
      data-popper-arrow=""
    />
  </div>
</transition-stub>
`;

exports[`BaseTooltip Snapshots matches Snapshot when position is right 1`] = `
<transition-stub
  enteractiveclass="enterActive"
  enterclass="enter"
  entertoclass="enterTo"
  leaveactiveclass="leaveActive"
  leaveclass="leave"
  leavetoclass="leaveTo"
>
  <div
    class="wrapper light"
    style="width: auto; height: auto;"
  >
    Example tooltip text 
    <div
      class="arrow"
      data-popper-arrow=""
    />
  </div>
</transition-stub>
`;

exports[`BaseTooltip Snapshots matches Snapshot when show is false 1`] = `
<transition-stub
  enteractiveclass="enterActive"
  enterclass="enter"
  entertoclass="enterTo"
  leaveactiveclass="leaveActive"
  leaveclass="leave"
  leavetoclass="leaveTo"
>
  <div
    class="wrapper light"
    style="width: auto; height: auto; display: none;"
  >
    Example tooltip text 
    <div
      class="arrow"
      data-popper-arrow=""
    />
  </div>
</transition-stub>
`;

exports[`BaseTooltip Snapshots matches Snapshot when theme is dark 1`] = `
<transition-stub
  enteractiveclass="enterActive"
  enterclass="enter"
  entertoclass="enterTo"
  leaveactiveclass="leaveActive"
  leaveclass="leave"
  leavetoclass="leaveTo"
>
  <div
    class="wrapper dark"
    style="width: auto; height: auto;"
  >
    Example tooltip text 
    <div
      class="arrow"
      data-popper-arrow=""
    />
  </div>
</transition-stub>
`;
